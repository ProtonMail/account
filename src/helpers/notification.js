import appDispatcher from 'frontend-commons/src/utils/appDispatcher';

const dispatcher = appDispatcher('notification');

const action = (type) => (message, opt = {}) => dispatcher.input(type, { message, opt });
export const success = action('success');
export const error = (err) => action('error')(err.message, err);
export const info = action('info');

export const onInput = dispatcher.onInput;
