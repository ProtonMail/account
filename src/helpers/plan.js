const searchParams = new URLSearchParams(window.location.search);

export const getDefaultConfigPlan = ({ Cycle = 12, Currency = 'EUR' } = {}) => ({
  Currency: searchParams.get('currency') || Currency,
  Cycle: +(searchParams.get('billing') || Cycle),
  Name: searchParams.get('plan') || 'vpnplus'
});

export const cycleString = ({ Cycle }, isMonth = false) => (Cycle === 12 && !isMonth) ? 'year' : 'month';
