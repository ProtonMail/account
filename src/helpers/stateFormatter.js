/**
 * Format the state and extend it as it's not recursive with unistore
 * @param  {Object} state
 * @param  {String} key
 * @param  {Object} value
 * @return {Object}       new state
 * @function
 */
export const toState = (state, key, value) => ({ [key]: { ...state[key], ...value } });

/**
 * Format the state and extends using the dot notation.
 * @param {Object} state - the state to update
 * @param {string|string[]} extendedKey - the key to update in the state. It can be either a string, either a list of string.
 * If it's a string, it corresponds to a path, the `.` separates the keys. If it's an array, each element is a key.
 * @param {Object} value - the updated value.
 * @returns {Object} new state.
 * @example
 * > extended(
 *     {
 *         a: 1,
 *         b: {
 *             c: {
 *                 d: 2,
 *                 e: 3
 *             },
 *             f: {
 *                 g: 4
 *             }
 *         }
 *     },
 *     'b.c',
 *     { d: 42, f: 25 }
 * );
 * // returns
 * {
 *     // a removed, because root keys are handled by unistore
 *     b: {
 *         c: {
 *             d: 42, // updated
 *             f: 25, // new
 *             e: 3   // unchanged
 *         },
 *         f: { // f is not removed, because subkeys are not handled by unistore
 *             g: 4
 *         }
 *     }
 * }
 *
 * @example
 * // beware!
 * > extended({...sameAsBefore }, ['b.c'], { d: 42, f: 25 })
 * { 'b.c': { d: 42, f: 25 } } // the new key 'b.c' is created
 * @function
 */
export const extended = (state, extendedKey, value) => {
    if (!Array.isArray(extendedKey)) {
        return extended(state, extendedKey.split('.'), value);
    }
    if (!extendedKey.length) {
        return value;
    }

    const [firstKey, ...rest] = extendedKey;
    return toState(state, firstKey, extended(state[firstKey], rest, value));
};
