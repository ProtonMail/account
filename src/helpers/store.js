import { createStore } from 'unistore/full/preact';
import devtools from 'unistore/devtools';

import authActions from '../actions/authentication';
import settingsActions from '../actions/settings';
import scopeActions from '../actions/scope';

export const initialState = {
    auth: {
        isLoggedIn: false,
        user: {},
        step: 'login',
        twoFactorResponse: {}
    },
    scope: {
        creds: null,
        response: null
    },
    settings: {
        addU2FKey: {},
        reset2FARecoveryCodes: {}
    }
};
const store = createStore(initialState);


export const actions = (store) => ({
    ...authActions(store),
    ...settingsActions(store),
    ...scopeActions(store)
});

export default store;
