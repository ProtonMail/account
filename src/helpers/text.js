import { saveAs } from 'file-saver';

/**
 * Download the given lines as a file.
 * @param {string} name - the name of the file
 * @param {string[]} lines - the lines to put in the file
 * @param {string} [type='text/plain;charset=utf-8'] - the mime type of the file
 */
export const downloadAsFile = (name, lines = [], type = 'text/plain;charset=utf-8') => {
    const blob = new Blob([lines.join('\r\n')], { type });
    saveAs(blob, name);
};
