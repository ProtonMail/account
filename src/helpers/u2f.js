import { sign, register } from 'u2f-api';
import appProvider from 'frontend-commons/src/appProvider';

export const ERROR_CODE = {
    SUCCESS: 0,
    OTHER_ERROR: 1,
    BAD_REQUEST: 2,
    CONFIGURATION_UNSUPPORTED: 3,
    DEVICE_INELIGIBLE: 4,
    TIMEOUT: 5
};

const ERROR_MAP = {
    [ERROR_CODE.SUCCESS]: () => {
        throw new Error('This is a success!');
    },
    [ERROR_CODE.OTHER_ERROR]: () => 'An error occurred',
    [ERROR_CODE.BAD_REQUEST]: () => 'An internal error occurred.',
    [ERROR_CODE.CONFIGURATION_UNSUPPORTED]: () => 'This security key is not supported.',
    [ERROR_CODE.DEVICE_INELIGIBLE]: (register) =>
        register ? 'This security key is already registered for your account!' : 'This security key is not recognized.',
    [ERROR_CODE.TIMEOUT]: () =>
        'Looks like you are taking too long to respond, please try again with a bit of motivation.'
};

/**
 * Sends a SIGN request to the U2F device
 * @param {Object} U2FRequest - the U2FRequest
 * @param {Object[]} U2FRequest.RegisteredKeys - the registered keys
 * @param {String} U2FRequest.Challenge - the challenge
 * @return {Promise<SignResponse>}
 */
export async function signU2F({ RegisteredKeys: registeredKeys, Challenge: challenge }) {
    const { appId, timeout } = appProvider.getConfig('u2f');

    const signRequest = registeredKeys.map(({ Version: version, KeyHandle: keyHandle }) => ({
        version,
        keyHandle,
        appId,
        challenge
    }));

    return await sign(signRequest, timeout);
}

/**
 * Sends a REGISTER request to the U2F device.
 * @param {Object} U2FRequest - the U2FRequest
 * @param {Object[]} U2FRequest.RegisteredKeys - the registered keys
 * @param {String} U2FRequest.Challenge - the challenge
 * @param {String[]} U2FRequest.Versions - the different versions accepted by the server
 * @return {Promise<RegisterResponse>}
 */
export async function registerU2F({ RegisteredKeys: registeredKeys, Challenge: challenge, Versions: versions }) {
    const { appId, timeout } = appProvider.getConfig('u2f');
    const signRequest = registeredKeys.map(({ Version: version, KeyHandle: keyHandle }) => ({
        version,
        keyHandle
    }));

    const registerRequests = versions.map((version) => ({
        version,
        challenge,
        appId
    }));

    return await register(registerRequests, signRequest, timeout);
}

/**
 * Computes an error message from a U2F error code.
 * @param {number} errorCode - an error code.
 * @param register
 * @return {*}
 */
export function getErrorMessage(errorCode, register = false) {
    if (ERROR_MAP[errorCode]) {
        return ERROR_MAP[errorCode](register);
    }
    throw new Error(`The error code "${errorCode}" does not exist`);
}
