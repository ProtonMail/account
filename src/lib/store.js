import { createStore } from 'unistore/full/preact';

import authActions from '../actions/authentication';
import settingsActions from '../actions/settings';

const store = createStore({
    auth: {
        isLoggedIn: false,
        user: {},
        step: 'login',
        twoFactorResponse: {}
    },
    settings: {
        addU2FKey: {},
        reset2FARecoveryCodes: {}
    }
});

export const actions = (store) => ({
    ...authActions(store),
    ...settingsActions(store)
});

export default store;
