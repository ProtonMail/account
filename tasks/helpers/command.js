const { promisify } = require('util');
const exec = promisify(require('child_process').exec);
const execRaw = require('child_process').exec;
const toCLI = (cmd) => (Array.isArray(cmd) ? cmd.join(' && ') : cmd);

const execVerbose = (command) => {
    return new Promise((resolve, reject) => {
        const callback = (error) => {
            if (error) {
                return reject(error);
            }
            resolve();
        };

        const build = execRaw(
            toCLI(command),
            {
                shell: '/bin/bash',
                maxBuffer: 1000 * 1000 * 10 // 10 MB
            },
            callback
        );
        build.stdout.pipe(process.stdout);
        build.stderr.pipe(process.stderr);
    });
};

module.exports = {
    execVerbose,
    exec(command) {
        return exec(toCLI(command), { shell: '/bin/bash' });
    }
};
