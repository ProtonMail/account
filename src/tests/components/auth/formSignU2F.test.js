import { isSupported } from 'u2f-api';

import ConnectedFormSignU2F, { FormSignU2F } from '../../../components/auth/formSignU2F';
import { ERROR_CODE } from '../../../helpers/u2f';
import store, { initialState } from '../../../helpers/store';
import { shallow } from 'preact-render-spy';
import { renderProvided, shallowProvider } from '../../testsHelpers/storeTools';

jest.mock('../../../actions/authentication');

describe('testing FormSignU2F component...', () => {
    afterEach(() => {
        store.setState(initialState, true);
    });

    test('error state', () => {
        const auth = {
            twoFactorResponse: {
                U2FResponse: {
                    metaData: {
                        code: 1
                    }
                }
            }
        };
        store.setState({ auth });
        expect(renderProvided(<ConnectedFormSignU2F />, store)).toMatchSnapshot();
    });

    test('U2F not available on this browser', () => {
        isSupported.mockReturnValueOnce(false);
        const auth = {
            twoFactorResponse: {}
        };
        store.setState({ auth });
        expect(renderProvided(<ConnectedFormSignU2F />, store)).toMatchSnapshot();
    });

    test('correct state', () => {
        const auth = {
            twoFactorResponse: {}
        };
        store.setState({ auth });
        expect(renderProvided(<ConnectedFormSignU2F />, store)).toMatchSnapshot();
    });

    test('success state', () => {
        const auth = {
            twoFactorResponse: { success: true }
        };
        store.setState({ auth });
        expect(renderProvided(<ConnectedFormSignU2F />, store)).toMatchSnapshot();
    });

    test('retry button', () => {
        const auth = {
            twoFactorResponse: {
                U2FResponse: {
                    metaData: {
                        code: ERROR_CODE.OTHER_ERROR
                    }
                }
            }
        };

        const loginU2FAction = jest.fn();
        const abortLoginAction = jest.fn();

        const component = shallow(
            <FormSignU2F auth={auth} loginU2FAction={loginU2FAction} abortLoginAction={abortLoginAction} />
        );

        component.find('[onClick]').simulate('click');
        expect(loginU2FAction).toHaveBeenCalledTimes(2);
        expect(abortLoginAction).toHaveBeenCalledTimes(0);
    });

    test('retry button when timeout', () => {
        const auth = {
            twoFactorResponse: {
                U2FResponse: {
                    metaData: {
                        code: ERROR_CODE.TIMEOUT
                    }
                }
            }
        };

        const loginU2FAction = jest.fn();
        const abortLoginAction = jest.fn();

        const component = shallow(
            <FormSignU2F auth={auth} loginU2FAction={loginU2FAction} abortLoginAction={abortLoginAction} />
        );

        component.find('[onClick]').simulate('click');
        expect(loginU2FAction).toHaveBeenCalledTimes(0);
        expect(abortLoginAction).toHaveBeenCalledTimes(2);
    });

    test('mount with empty state', () => {
        const auth = {
            twoFactorResponse: {
                success: false
            }
        };

        const loginU2FAction = jest.fn();
        const abortLoginAction = jest.fn();

        const component = shallow(
            <FormSignU2F auth={auth} loginU2FAction={loginU2FAction} abortLoginAction={abortLoginAction} />
        );

        expect(loginU2FAction).toHaveBeenCalledTimes(1);
        expect(abortLoginAction).toHaveBeenCalledTimes(0);
    });
});
