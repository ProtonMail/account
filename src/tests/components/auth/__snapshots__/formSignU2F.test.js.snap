// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing FormSignU2F component... U2F not available on this browser 1`] = `
<div>
  <p>Your browser is not supported, please use another 2FA method instead</p>
</div>
`;

exports[`testing FormSignU2F component... correct state 1`] = `
<div>
  <p>Activate your security key...</p>
</div>
`;

exports[`testing FormSignU2F component... error state 1`] = `
<div>
  <p>Activate your security key...</p>
  <p>
    <span>An error occurred.</span>
    <button type="button">Retry</button>
  </p>
</div>
`;

exports[`testing FormSignU2F component... success state 1`] = `
<div>
  <p>Success</p>
</div>
`;
